### PR TITLE
Migrate min(dim=?) from THC to ATen and remove _min

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Related issue: #36900

Since I feel this PR is already large enough, I didn't migrate max in this PR. Legacy code is not cleaned up either. All these remaining work will be done in later PRs after this is merged.

Benchmark on an extreme case
```python
import torch
print(torch.__version__)

t = torch.randn(100000, 2, device='cuda')

warmup = torch.arange(100000000)
torch.cuda.synchronize()

%timeit t.min(dim=0); torch.cuda.synchronize()
```
Before: 4ms; After: 24.5us.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
@ngimel Finally, ROCm bug is fixed!!!

**Comment by zasdfgbnm** [comment]
Just resolved a conflict

**Comment by ngimel** [comment]
bc-breaking, as it removes _min function from native_functions.yaml. This was an internal function that is not documented or used in user scripts, it is also not traced/scripted. Model compatibility test passes.

File: aten/src/ATen/cuda/DeviceUtils.cuh [hunk 1]
```diff
@@ -58,15 +58,33 @@ __device__ __forceinline__ T WARP_SHFL_UP(T value, unsigned int delta, int width
...
+template<>
+__device__ __forceinline__ c10::Half WARP_SHFL_DOWN<c10::Half>(c10::Half value, unsigned int delta, int width, unsigned int mask)
+{
+  //(HIP doesn't support c10::Half). Trick from https://devblogs.nvidia.com/faster-parallel-reductions-kepler/
+  short a = *reinterpret_cast<short*>(&value);
+  a = __shfl_down(a, delta);
+  return *reinterpret_cast<c10::Half*>(&a);
+}
+template<>
+__device__ __forceinline__ __half WARP_SHFL_DOWN<__half>(__half value, unsigned int delta, int width, unsigned int mask)
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/cuda/DeviceUtils.cuh:79]: I don't know, I am tired of debugging it and added both, and it works. I don't know which one makes it work. 😰

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/cuda/DeviceUtils.cuh:79]: The message was `__shfl_down` not defined for `__half`

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply aten/src/ATen/cuda/DeviceUtils.cuh:72]: I refactored it, now `c10::Half` specialization is handled completely `DeviceUtils.cuh`, and there is no more `wrap_half` in `SharedReduceOps.h`

**Review by ngimel — APPROVED** [review]
